DataColumnSidecars Sampler should sample minimal 8 columns not 4

jeroost opened the pull request:
Per spec:
At each slot, a node advertising custody_group_count downloads a minimum of sampling_size = max(SAMPLES_PER_SLOT, custody_group_count) custody groups, selected by groups = get_custody_groups(node_id, sampling_size), to which correspond the columns compute_columns_for_custody_group(group) for group in groups

<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/Consensys/teku/blob/master/CONTRIBUTING.md -->

## PR Description

## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->
fixes #9598 
## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

jeroost committed:
DataColumnSidecars Sampler should sample minimal 8 columns not 4
At each slot, a node advertising custody_group_count downloads a minimum of sampling_size = max(SAMPLES_PER_SLOT, custody_group_count) custody groups, selected by groups = get_custody_groups(node_id, sampling_size), to which correspond the columns compute_columns_for_custody_group(group) for group in groups

jeroost committed:
DasSamplerBasic - Adding tests

zilm13 commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/CustodyGroupCountManagerImpl.java:
```diff
@@ -131,6 +131,12 @@ public List<UInt64> getCustodyColumnIndices() {
 
+  @Override
+  public List<UInt64> getSamplingColumnIndices() {
+    final int sampleCount = Math.max(getCustodyGroupCount(), specConfigFulu.getSamplesPerSlot());
```
Let's call it  sampleGroupCount for clarity

jeroost replied:
done

zilm13 commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/CustodyGroupCountManagerImpl.java:
```diff
@@ -131,6 +131,12 @@ public List<UInt64> getCustodyColumnIndices() {
   }
 
+  @Override
+  public List<UInt64> getSamplingColumnIndices() {
```
Could you plese also add small unit test on this method that sampling columns include all custody columns with default 4/8 init?

jeroost replied:
done, see CustodyGroupCountManagerImplTest

zilm13 commented on ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerBasicTest.java:
```diff
@@ -0,0 +1,184 @@
+              eq(new DataColumnSlotAndIdentifier(UInt64.ZERO, blockRoot, columnIndex))))
+          .thenReturn(SafeFuture.completedFuture(true));
+    }
+    final List<UInt64> nonCustodyColumns = new ArrayList<>();
```
Maybe samplingNonCustodyColumns? or samplingOnlyColumns. `nonCustodyColumns` is a bit confusing

jeroost replied:
agreed, done

zilm13 commented on ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerBasicTest.java:
```diff
@@ -0,0 +1,184 @@
+      when(custody.onNewValidatedDataColumnSidecar(any())).thenReturn(SafeFuture.COMPLETE);
+    }
+
+    SafeFuture<List<UInt64>> result = sampler.checkDataAvailability(UInt64.ZERO, blockRoot);
```
finals please here and there

jeroost replied:
done
Wish there was an IDE default setting for that...

zilm13 commented on ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerBasicTest.java:
```diff
@@ -0,0 +1,184 @@
+
+    // Add assertions
+    assertThat(availableColumns).containsAll(custodyGroupCountManager.getCustodyColumnIndices());
+    assertThat(availableColumns).containsAll(custodyGroupCountManager.getSamplingColumnIndices());
```
Please use`containsExactlyInAnyOrder` here or smth like this  to clarify that it's ==

jeroost replied:
done, for the latter

zilm13 reviewed (commented):
LGTM just some nits, but I'd add simple test for clarification

jeroost committed:
Add new unit test and process PR feedback

jeroost committed:
Add new unit test and process PR feedback

zilm13 committed:
Merge branch 'master' into das-sampler-min-8-columns

zilm13 committed:
Add gossip subscription for samples per slot

jeroost committed:
DataColumnSidecarSubnetBackboneSubscriber should look at samplingGroupCount, not custodyGroupCount. If not, it may not be able to retrieve the correct columns because there are not the right peers connected.

jeroost committed:
Merge remote-tracking branch 'jeroost/das-sampler-min-8-columns' into das-sampler-min-8-columns

# Conflicts:
#	services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java

jeroost committed:
Fix #9622 investigate on ExtraDataAppendedException

DataColumnSideCar maximumResponseChunks should take the maximum batch size into account. If not, Eth2OutgoingRequestHandler will terminate because chunksReceived == maximumResponseChunks and the connection terminates.

jeroost committed:
Rebase to master

jeroost committed:
Merge branch 'master' into das-sampler-min-8-columns

jeroost committed:
Fix merge from master

jeroost committed:
Cleanup and undo other deltas

jeroost committed:
Cleanup pre PR

zilm13 commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/CustodyGroupCountManager.java:
```diff
@@ -42,6 +52,10 @@ public void setCustodyGroupSyncedCount(int custodyGroupSyncedCount) {}
 
   List<UInt64> getCustodyColumnIndices();
 
+  int getSampleGroupCount();
```
nit: maybe `getSamplingGroupCount` in line with other methods name?

jeroost committed:
Merge branch 'master' into das-sampler-min-8-columns

zilm13 commented on ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodyStand.java:
```diff
@@ -75,6 +77,8 @@ private DasCustodyStand(
       final int totalCustodyGroupCount,
       final Optional<Duration> asyncDbDelay,
       final Optional<Duration> asyncBlockResolverDelay) {
+    final SpecVersion specVersion = spec.forMilestone(SpecMilestone.FULU);
```
our purpose for the Stand is to test, so we'd better to add sampling input rather than calculate something that cannot be mocked or overridden

zilm13 commented on ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/CustodyGroupCountManagerImplTest.java:
```diff
@@ -0,0 +1,77 @@
+    final List<UInt64> samplingColumnIndices = custodyGroupCountManager.getSamplingColumnIndices();
+    // Sampling column groups should always include all custody columns at the minimum.
+    assertThat(samplingColumnIndices)
+        .contains(custodyGroupCountManager.getCustodyColumnIndices().toArray(new UInt64[] {}));
```
`containsAll` will do the thing w/o intermediate array

zilm13 commented on ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/CustodyGroupCountManagerImplTest.java:
```diff
@@ -0,0 +1,77 @@
+                            .dataColumnSidecarSubnetCount(4)
+                            .numberOfColumns(128)
+                            .numberOfCustodyGroups(128)
+                            .custodyRequirement(4)
```
please put 4 and 8 in constants and use them appropriately

zilm13 commented on ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerBasicTest.java:
```diff
@@ -0,0 +1,186 @@
+    // Add assertions
+    assertThat(availableColumns).containsAll(custodyGroupCountManager.getCustodyColumnIndices());
+    assertThat(availableColumns)
+        .containsExactlyInAnyOrder(
```
containsExactlyInAnyOrderElementsOf

zilm13 reviewed (commented):
LGTM, just few nits

jeroost committed:
Processing review comments

jeroost committed:
Post merge compilation issue

jeroost committed:
Cleanup last nits

jeroost committed:
Cleanup last nits

zilm13 reviewed (approved):
Thank you for your work, looks perfect, merging
